pkg/filters/opafilter/opafilter.go

```diff
@@ -0,0 +1,241 @@
+type RegoResult struct {
+	Allow             bool              `json:"allow"`
+	AdditionalHeaders map[string]string `json:"additional_headers,omitempty"`
+	StatusCode        int               `json:"status_code,omitempty"`
+}
+
+var kind = &filters.Kind{
+	Name:        kindName,
+	Description: "OPAFilter implement OpenPolicyAgent function",
+	Results:     []string{resultFiltered},
+	DefaultSpec: func() filters.Spec {
+		return &Spec{
+			DefaultStatus: 403,
+		}
+	},
+	CreateInstance: func(spec filters.Spec) filters.Filter {
+		return &OPAFilter{spec: spec.(*Spec)}
+	},
+}
+
+type OPAFilter struct {
+	spec                  *Spec
+	includedHeadersParsed []string `yaml:"includedHeadersParsed"`
+	regoQuery             *rego.PreparedEvalQuery
+}
+
+type Spec struct {
+	filters.BaseSpec `yaml:",inline"`
+	DefaultStatus    int    `yaml:"defaultStatus"`
+	IncludedHeaders  string `yaml:"includedHeaders"`
+	ReadBody         bool   `yaml:"readBody"`
+	Policy           string `yaml:"policy"`
+}
+
+func init() {
+	filters.Register(kind)
+}
+
+func (o *OPAFilter) Name() string {
+	return o.spec.Name()
+}
+
+func (o *OPAFilter) Spec() filters.Spec {
+	return o.spec
+}
+
+func (o *OPAFilter) Kind() *filters.Kind {
+	return kind
+}
+
+func (o *OPAFilter) Init() {
+	o.includedHeadersParsed = strings.Split(o.spec.IncludedHeaders, ",")
+	if o.spec.DefaultStatus == 0 {
+		o.spec.DefaultStatus = 403
+	}
+	n := 0
+	for i := range o.includedHeadersParsed {
+		scrubbed := strings.ReplaceAll(o.includedHeadersParsed[i], " ", "")
+		if scrubbed != "" {
+			o.includedHeadersParsed[n] = textproto.CanonicalMIMEHeaderKey(scrubbed)
+			n++
+		}
+	}
+	o.includedHeadersParsed = o.includedHeadersParsed[:n]
+	ctx, cancelFunc := stdctx.WithTimeout(stdctx.Background(), 5*time.Second)
+	query, err := rego.New(
+		rego.Query("result = data.http.allow"),
+		rego.Module("inline.rego.policy", o.spec.Policy),
+	).PrepareForEval(ctx)
+	defer cancelFunc()
+	if err != nil {
+		logger.Errorf("create PrepareForEval rego query error: %s", err)
+	} else {
+		o.regoQuery = &query
+	}
+}
+
+func (o *OPAFilter) Inherit(previousGeneration filters.Filter) {
+	o.Init()
+	previousGeneration.Close()
+}
+
+func (o *OPAFilter) Handle(ctx *context.Context) (result string) {
+	req := ctx.GetInputRequest().(*httpprot.Request)
+	var rw *httpprot.Response
+	if rw, _ = ctx.GetOutputResponse().(*httpprot.Response); rw == nil {
+		rw, _ = httpprot.NewResponse(nil)
+		ctx.SetOutputResponse(rw)
+	}
+	return o.evalRequest(req, rw)
+}
+
+func (o *OPAFilter) Status() interface{} {
+	return nil
+}
+
+func (o *OPAFilter) Close() {
+}
+
+func (o *OPAFilter) evalRequest(r *httpprot.Request, w *httpprot.Response) string {
+	headers := map[string]string{}
+
+	for key, value := range r.HTTPHeader() {
+		if len(value) > 0 && stringtool.StrInSlice(key, o.includedHeadersParsed) {
+			headers[key] = strings.Join(value, ", ")
+		}
+	}
+
+	var body string
+	if o.spec.ReadBody {
+		buf, _ := io.ReadAll(r.Body)
```

Comment:
please use the payload functions to access the body, refer:
https://github.com/megaease/easegress/blob/main/pkg/protocols/httpprot/request.go#L40-L42

and we may also need to handle stream & non-stream bodies separately, refer:
https://github.com/megaease/easegress/blob/main/pkg/filters/requestadaptor/requestadaptor.go#L308-L322